### ISSUE TITLE
Check weights for NANs

### ISSUE THREAD

**Issue by r-luo** [microsoft/LightGBM#2359]
## Environment info

Operating System: AWS RHEL 7

CPU/GPU model: AWS ec2 m5.24xlarge

C++/Python/R version: Python 3.7

LightGBM version or commit hash: 2.2.3

### Set up variables
```python
# Set up datasets
x1 = pd.DataFrame(np.random.random(size=(100, 2))).values
y1 = np.log(x1[:, 0] ** 2 + x1[:, 1] ** 0.5) + np.random.normal(size=100) > 0
x2 = pd.DataFrame(np.random.random(size=(100, 2))).values
y2 = np.log(x2[:, 0] ** 2 + x2[:, 1] ** 0.5) + np.random.normal(size=100) > 0
w1 = np.random.random(size=100)
w2 = np.random.random(size=100)

param_dict = {
    'objective': 'binary',
     'tree_learner': 'feature',
     'metric': ['auc'],
     'learning_rate': 0.1,
     'max_bin': 127,
     'max_depth': 10,
     'num_leaves': 20,
     'min_data_in_leaf': 2,
     'num_threads': 48
}
```

### Training without weights
```python
# Training without weights
lgb_train = lgb.Dataset(x1, y1, free_raw_data=False)
lgb_val = lgb.Dataset(x2, y2, free_raw_data=False)

evals_result = {}
m_nw = lgb.train(
    params=param_dict, 
    train_set=lgb_train, 
    valid_sets=(lgb_val,), 
    num_boost_round=100, 
    early_stopping_rounds=10, 
    evals_result=evals_result
)
```
Returns
```
[1]	valid_0's auc: 0.584848
Training until validation scores don't improve for 5 rounds.
[2]	valid_0's auc: 0.571717
[3]	valid_0's auc: 0.564242
[4]	valid_0's auc: 0.564646
[5]	valid_0's auc: 0.589495
[6]	valid_0's auc: 0.582424
[7]	valid_0's auc: 0.582828
[8]	valid_0's auc: 0.571515
[9]	valid_0's auc: 0.566465
[10]	valid_0's auc: 0.588283
Early stopping, best iteration is:
[5]	valid_0's auc: 0.589495
```

### Training with weights

```python
# Training with weights
lgb_train = lgb.Dataset(x1, y1, weight=w1, free_raw_data=False)
lgb_val = lgb.Dataset(x2, y2, weight=w2, free_raw_data=False)

evals_result = {}
m_w = lgb.train(
    params=params,
    train_set=lgb_train, 
    valid_sets=(lgb_val,), 
    num_boost_round=100, 
    early_stopping_rounds=5, 
    evals_result=evals_result
)
```
returns 
```
[1]	valid_0's auc: 0.5
Training until validation scores don't improve for 5 rounds.
[2]	valid_0's auc: 0.5
[3]	valid_0's auc: 0.5
[4]	valid_0's auc: 0.5
[5]	valid_0's auc: 0.5
[6]	valid_0's auc: 0.5
Early stopping, best iteration is:
[1]	valid_0's auc: 0.5
```

In addition, the weight usage example here https://github.com/microsoft/LightGBM/blob/master/examples/python-guide/advanced_example.py#L18-L19 only used a constant weight of `1.0`, which is reset to `None` when the `lgb.Dataset` is initialized (https://github.com/microsoft/LightGBM/blob/master/python-package/lightgbm/basic.py#L1356-L1357)

**Comment by StrikerRUS** [comment]
@r-luo I think you should use something smarter than random weights to see AUC improvement. However, even with your example I get the following output:
```
[1]	valid_0's auc: 0.515746
Training until validation scores don't improve for 10 rounds.
[2]	valid_0's auc: 0.511844
[3]	valid_0's auc: 0.516209
[4]	valid_0's auc: 0.531094
[5]	valid_0's auc: 0.538052
[6]	valid_0's auc: 0.534307
[7]	valid_0's auc: 0.555308
[8]	valid_0's auc: 0.549065
[9]	valid_0's auc: 0.536582
[10]	valid_0's auc: 0.54061
[11]	valid_0's auc: 0.543462
[12]	valid_0's auc: 0.551296
[13]	valid_0's auc: 0.551614
[14]	valid_0's auc: 0.545843
[15]	valid_0's auc: 0.546822
[16]	valid_0's auc: 0.551793
[17]	valid_0's auc: 0.548864
Early stopping, best iteration is:
[7]	valid_0's auc: 0.555308
```
Use `np.random.seed(42)` to get the same output.

**Comment by StrikerRUS** [comment]
@guolinke What do you think about ["all-ones" weights](https://github.com/microsoft/LightGBM/blob/master/examples/binary_classification/binary.test.weight) in our example?

**Comment by guolinke** [comment]
@r-luo thanks, I think we should throw an error when meeting `nan`.

**Comment by StrikerRUS** [comment]
@guolinke Is it job for cpp code?

**Comment by guolinke** [comment]
Yes, i will do it when have time
